src/com/aaronzalewski/HtmlDocumentImpl.java

```diff
@@ -0,0 +1,85 @@
+package com.aaronzalewski;
+
+import com.google.common.collect.ImmutableList;
+import org.jsoup.Connection;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class HtmlDocumentImpl implements HtmlDocument {
+
+    private static final String EMAIL_REGEX = "[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+";
+
+    private String domain;
+    private Document document;
+
+    public static HtmlDocument createHtmlDocument(String url) {
+        try {
+            return new HtmlDocumentImpl(url);
+        } catch (IOException | IllegalArgumentException e) {
+            return new EmptyHtmlDocumentImpl();
+        }
+    }
+
+    private HtmlDocumentImpl(String url) throws IOException {
+        Connection connection = Jsoup.connect(url);
+        this.document = connection.get();
+        this.domain = getUrlDomain(url);
+    }
+
+    @Override
+    public List<String> getEmailAddresses() {
+        ImmutableList.Builder<String> emails = ImmutableList.builder();
+
+        Matcher matcher = Pattern.compile(EMAIL_REGEX).matcher(document.text());
+        while(matcher.find()) {
+            emails.add(matcher.group());
+        }
+
+        return emails.build();
+    }
+
+    @Override
+    public List<String> getUrlsInDomain() {
+        ImmutableList.Builder<String> links = ImmutableList.builder();
+        document.select("a[href*=" + domain + "]").stream()
+                .filter(link -> (hasDocumentDomain(link)))
+                .forEach(link -> links.add(link.attr("href")));
+
+        return links.build();
+    }
+
+    private String getUrlDomain(String url) {
+        try {
+            return (new URI(url)).getHost();
+        } catch (URISyntaxException e) {
+            return "";
+        }
+    }
+
+    private boolean hasDocumentDomain(Element link) {
+        return domain.equals(getUrlDomain(link.attr("href"))) ||
+                ("www." + domain).equals(getUrlDomain(link.attr("href")));
```

Comment:
I decided to check for www in the case that someone inputs a domain such as jana.com.  In that case, it directs to the equivalent page as www.jana.com, but all the links on the page inlcude www.jana.com, so a test on the domain would fail in that case without checking for www.  I check for www specifically because www is the most common case for this.

Should I restrict it so that the user would have to input www.jana.com in the above example?  Or should I include all subdomains of the given domain? Or should I accept subdomains that direct to the same page as the parent domain?
I tried implementing the third option to allow the subdomain www, which resulted in what you see above.  I agree that this would result in allowing something like www.web.mit.edu, which doesn't make much sense.